include/dynamatic/Analysis/ControlDependenceAnalysis.h

```diff
@@ -0,0 +1,138 @@
+  ControlDependenceAnalysis(Operation *operation) {
+    // type-cast it into mlir::func::FuncOp
+    mlir::func::FuncOp funcOp = dyn_cast<mlir::func::FuncOp>(operation);
+    if (funcOp) {
+      identifyAllControlDeps(funcOp);
+    } else {
+      // type-cast it into mlir::ModuleOp
+      ModuleOp modOp = dyn_cast<ModuleOp>(operation);
+      if (modOp) {
+        auto funcOps = modOp.getOps<mlir::func::FuncOp>();
+        // call those in a loop and create a big structure
+        for (mlir::func::FuncOp funcOp : llvm::make_early_inc_range(funcOps)) {
+          identifyAllControlDeps(funcOp);
+        }
+      } else
+        // report an error indicating that the anaylsis is instantiated over
+        // an inappropriate operation
+        llvm::errs() << "ControlDependenceAnalysis is instantiated over an "
+                        "operation that is not FuncOp or ModuleOp!\n";
+    }
+  };
+
+  // All public functions are accessible from outside of the pass and require
+  // the outside to specify the index of the funcOp of interest. If the outside
+  // has a func::FuncOp rather than a ModuleOp, they should just pass a 0
+
+  // return all BBs that the block in the argument, belonging to the funcOp of
+  // the passed idx, is control dependent on
+  void
+  calculateBlockControlDeps(Block *block, int funcOp_idx,
+                            SmallVector<Block *, 4> &returned_control_deps);
+
+  // return only forward dependencies (i.e., excluding loop exits) that the
+  // block in the argument, belonging to the funcOp of the passed idx, is
+  // control dependent on
+  void calculateBlockForwardControlDeps(
+      Block *block, int funcOp_idx,
+      SmallVector<Block *, 4> &returned_forward_control_deps);
+
+  // loops over the blocks and prints to the terminal the block's control
+  // dependencies of the passed function
+  void printAllBlocksDeps(int funcOp_idx);
+  void printForwardBlocksDeps(int funcOp_idx);
+
+  /// Invalidation hook to keep the analysis cached across passes. Returns true
+  /// if the analysis should be invalidated and fully reconstructed the next
+  /// time it is queried.
+  bool isInvalidated(const mlir::AnalysisManager::PreservedAnalyses &pa) {
+    return !pa.isPreserved<ControlDependenceAnalysis>();
+  }
+
+private:
+  // For every func::FuncOp, every basic block from the CFG is mapped to a
+  // vector containing all basic blocks that it is control dependent on We are
+  // using SmallVector because it is more efficient than std::vector which use
+  // heap allocations If the content of SmallVec tor exceeds 4, it will still be
+  // functional but will start using the heap
+  SmallVector<DenseMap<Block *, SmallVector<Block *, 4>>, 4>
+      all_control_deps_maps;
+
+  // For every func::FuncOp, contains only the forward control dependencies,
+  // excluding the loop exit conditions
+  SmallVector<DenseMap<Block *, SmallVector<Block *, 4>>, 4>
+      forward_control_deps_maps;
```

Comment:
This two data types are complex and, while to the reader they seem disjoint, they are actually semantically linked (corresponding entries in both vectors represent the same function) which is slightly confusing. It also is unclear to me to use function iteration order to associate entries to functions, I think that a much more natural representation would be to map each function to its data. I propose the following organization instead (boilerplate names, modify as appropriate), which brings some more structure to the data.

```cpp

struct BlockInfo {
  /// Here I am assuming that a set is more appropriate than a vector
  /// as I suggested in other in-line remarks.
  DenseSet<Block*> controlDeps;
  DenseSet<Block*> forwardControlDeps;
};

using BlockMap = DenseMap<Block*, BlockInfo>;

DenseMap<func::FuncOp, BlockMap> funcDependencies;
```